### PR TITLE
Cancel On-going Uploads

### PR DESCRIPTION
**Pull request by deepakkinni**
1. Introduced new field "UploadAbort" in UploadSpec to indicate a request for upload abort.
2. On delete snapshot, first, a check is made to see if there is an ongoing upload.
3. If there is one, the cancellation is triggered.
4. Introduced a new phase called UploadPhaseCanceled, this indicates the upload was prematurely canceled.
5. formatted code using go fmt.

### CODE REVIEW

File: pkg/snapshotmgr/snapshot_manager.go [hunk 1]
```diff
@@ -183,30 +187,95 @@ func (this *SnapshotManager) CreateSnapshot(peID astrolabe.ProtectedEntityID, ta
...
+	uploadName := "upload-" + peID.GetSnapshotID().GetID()
+	log.Infof("Searching for Upload CR: %v", uploadName)
+	uploadCR, err := pluginClient.VeleropluginV1().Uploads(veleroNs).Get(uploadName, metav1.GetOptions{})
 	if err != nil {
-		if !strings.Contains(err.Error(), "The specified key does not exist") {
-			log.WithError(err).Errorf("Failed to delete the durable snapshot for PEID")
+		log.WithError(err).Errorf(" Error while retrieving the upload CR %v", uploadName)
+	}
+	uploadCompleted := uploadCR.Status.Phase == v1api.UploadPhaseCompleted
+	if uploadCR != nil && !uploadCompleted {
```
Comment (lintongj) [inline pkg/snapshotmgr/snapshot_manager.go:215]: Maybe we can reuse the existing code of patching Upload CR available in upload controller.

File: pkg/snapshotmgr/snapshot_manager.go [hunk 2]
```diff
@@ -183,30 +187,95 @@ func (this *SnapshotManager) CreateSnapshot(peID astrolabe.ProtectedEntityID, ta
...
+	}
+	uploadName := "upload-" + peID.GetSnapshotID().GetID()
+	log.Infof("Searching for Upload CR: %v", uploadName)
+	uploadCR, err := pluginClient.VeleropluginV1().Uploads(veleroNs).Get(uploadName, metav1.GetOptions{})
 	if err != nil {
-		if !strings.Contains(err.Error(), "The specified key does not exist") {
-			log.WithError(err).Errorf("Failed to delete the durable snapshot for PEID")
+		log.WithError(err).Errorf(" Error while retrieving the upload CR %v", uploadName)
+	}
+	uploadCompleted := uploadCR.Status.Phase == v1api.UploadPhaseCompleted
```
Comment (lintongj) [inline pkg/snapshotmgr/snapshot_manager.go:214]: If it is in the state of UploadPhaseCleanupFailed, do we also submit an upload cancel request?

File: pkg/controller/upload_controller.go [hunk 3]
```diff
@@ -294,14 +338,17 @@ func (c *uploadController) processUpload(req *pluginv1api.Upload) error {
 		return errors.New(errMsg)
 	}
 
+	// Unregister on-going upload
+	c.dataMover.UnregisterOngoingUpload(peID)
```
Comment (lintongj) [inline pkg/controller/upload_controller.go:342]: would it be more accurate to do the unregister right after upload is completed, compared to do it after the cleanup of local snapshot?

File: pkg/controller/upload_controller.go [hunk 4]
```diff
@@ -272,13 +303,26 @@ func (c *uploadController) processUpload(req *pluginv1api.Upload) error {
...
 	_, err = c.dataMover.CopyToRepo(peID)
 	if err != nil {
-		errMsg := fmt.Sprintf("Failed to upload snapshot, %v, to durable object storage. %v", peID.String(), errors.WithStack(err))
-		_, err = c.patchUploadByStatus(req, pluginv1api.UploadPhaseUploadError, errMsg)
-		if err != nil {
-			errMsg = fmt.Sprintf("%v. %v", errMsg, errors.WithStack(err))
+		log.Infof("CopyToRepo Error Received: %v", err.Error())
+		// Check if the request was canceled.
+		// context canceled error is wrapped in awserr, hence a string compare.
+		if strings.Contains(err.Error(), context.Canceled.Error()) {
```
Comment (lintongj) [inline pkg/controller/upload_controller.go:309]: It is not recommended to do error type checking using string comparison. The following reference might be useful.

https://stackoverflow.com/questions/47179024/how-to-check-if-a-request-was-cancelled

Comment (deepakkinni) [reply pkg/controller/upload_controller.go:309]: I agree, but s3 APIs do something weird with the RequestCancelled error, they wrap it up in their own "aswerr" type. I assumed doing errors.Is(err, context.Canceled) would work, but that doesn't either. The error string compare was a last resort.

File: pkg/snapshotmgr/snapshot_manager.go (hunk 2 above)
Comment (deepakkinni) [reply pkg/snapshotmgr/snapshot_manager.go:214]: Interesting, UploadPhaseCleanupFailed will only occur after the upload was completed and it failed to cleanup local snapshots, this means, UploadPhaseCleanupFailed should be an equivalent to UploadPhaseCompleted. Essentially, as long as we are not in a terminal state, i.e, Completed, CleanupFailed, Canceled, UploadError, then we should cancel. Will make these changes.

File: pkg/controller/upload_controller.go [hunk 5]
```diff
@@ -103,17 +104,30 @@ func (c *uploadController) enqueueUploadItem(obj interface{}) {
 	switch req.Status.Phase {
 	case "", pluginv1api.UploadPhaseNew, pluginv1api.UploadPhaseInProgress, pluginv1api.UploadPhaseUploadError:
 		// Process New and InProgress and UploadError Uploads
+	case pluginv1api.UploadPhaseCanceled:
```
Comment (swgupta-vmw) [inline pkg/controller/upload_controller.go:107]: Move this to the top. Do we want the UploadPhaseNew and others to fall through to the default case?

File: pkg/controller/upload_controller.go [hunk 6]
```diff
@@ -172,10 +186,22 @@ func (c *uploadController) processUploadItem(key string) error {
 		// For UploadPhaseInProgress, the resource lease logic will process the Upload if the lease is not held by
 		// another DataManager. If the DataManager holding the lease has died and/or lease has expired the current node
 		// will pick such record in UploadPhaseInProgress status for processing.
+	case pluginv1api.UploadPhaseCanceled:
```
Comment (swgupta-vmw) [inline pkg/controller/upload_controller.go:189]: Same as before. Move up?

File: pkg/controller/upload_controller.go (hunk 5 above)
Comment (deepakkinni) [reply pkg/controller/upload_controller.go:107]: golang doesn't have fallthrough for switch statement.

File: pkg/apis/veleroplugin/v1/upload.go [hunk 7]
```diff
@@ -28,18 +28,22 @@ type UploadSpec struct {
...
 type UploadPhase string
 
 const (
 	UploadPhaseNew           UploadPhase = "New"
 	UploadPhaseInProgress    UploadPhase = "InProgress"
 	UploadPhaseCompleted     UploadPhase = "Completed"
-	UploadPhaseUploadError  UploadPhase = "UploadError"
+	UploadPhaseUploadError   UploadPhase = "UploadError"
 	UploadPhaseCleanupFailed UploadPhase = "CleanupFailed"
+	UploadPhaseCanceled 	 UploadPhase = "Canceled"
```
[the remark this answers is not in the document]
Comment (deepakkinni) [reply pkg/apis/veleroplugin/v1/upload.go:46]: i've changed it to canceled.

File: pkg/controller/upload_controller.go (hunk 6 above)
Comment (deepakkinni) [reply pkg/controller/upload_controller.go:189]: golang doesn't have fallthrough for switch statement.

File: pkg/apis/veleroplugin/v1/upload.go [hunk 8]
```diff
@@ -28,18 +28,22 @@ type UploadSpec struct {
...
```
Comment (dsu-igeek) [inline pkg/apis/veleroplugin/v1/upload.go:46]: I think we should have a Canceling phase as well.  Once you set the flag, you don't know if the controller is reacting to it or not.

**Comment by lintongj** [comment]
Hi @deepakkinni, please add a changelog for this change. Please follow files in https://github.com/vmware-tanzu/velero-plugin-for-vsphere/tree/master/changelogs/unreleased as examples. Thanks.

File: pkg/snapshotmgr/snapshot_manager.go [hunk 9]
```diff
@@ -183,30 +184,83 @@ func (this *SnapshotManager) CreateSnapshot(peID astrolabe.ProtectedEntityID, ta
...
+		log.WithError(err).Errorf(" Error while retrieving the upload CR %v", uploadName)
+	}
+	// An upload is considered done when it's in either of the terminal stages- Completed, CleanupFailed, Canceled
+	uploadCompleted := this.isTerminalState(uploadCR)
+	if uploadCR != nil && !uploadCompleted {
+		log.Infof("Found the Upload CR: %v, updating spec to indicate cancel upload.", uploadName)
+		timeNow := clock.RealClock{}
+		mutate := func(r *v1api.Upload) {
+			uploadCR.Spec.UploadCancel = true
+			uploadCR.Status.Phase = v1api.UploadPhaseCanceling
```
Comment (dsu-igeek) [inline pkg/snapshotmgr/snapshot_manager.go:218]: Let's make the canceling phase be what the data manager sets when it starts processing the cancel.  I want to be able to see if the data manager is paying attention to the cancel request.  If it never moves to UploadPhaseCanceling then we know that the Data Manager is having a problem seeing the CR.  If it gets stuck in UploadPhaseCanceling then we know it has a problem stopping the upload.

**Pull request closed without merging**